### DISCUSSION TITLE
Saving predictions correctly

### DISCUSSION THREAD

**Discussion by lcoandrade** [azavea/raster-vision#1810] [Q&A] [answered]
I was able to make my first complete semantic segmentation pipeline using RV as a module. I made the code on this Kaggle [notebook](https://www.kaggle.com/code/luizclaudioandrade/rastervision-101) with this [dataset](https://www.kaggle.com/datasets/luizclaudioandrade/naip-chesapeake-sample).

The predictions are ok for the validation dataset (considering the number os epochs used), but when I use my learner to generate de predictions for my prediction dataset:
```
from rastervision.core.data import SemanticSegmentationLabels

predictions = learner.predict_dataset(
    dataset=prediction_dataset,
    raw_out=True,
    numpy_out=True,
    predict_kw=dict(out_shape=IMG_SIZE), # using 256x256
    progress_bar=True,
)

pred_labels = SemanticSegmentationLabels.from_predictions(
    windows=prediction_dataset.windows,
    predictions=predictions,
    smooth=True, # trying both outputs
    extent=prediction_dataset.scene.extent,
    num_classes=len(class_config),
)
```

The results are not good. What am I doing wrong?

Here are the predictions for my validation dataset:
<img width="967" alt="Captura de Tela 2023-06-24 às 15 31 54" src="https://github.com/azavea/raster-vision/assets/8769816/93189b88-ce73-40a4-8235-f62188ec82fc">

And here are the geotif file saved with:
```
pred_labels.save(
    uri=os.path.join(OUTPUT_DIR, 'predictions'),
    crs_transformer=prediction_dataset.scene.raster_source.crs_transformer,
    class_config=class_config,
)
```

On the left we have the labels for the window displayed and on the right we have the same window on the geotiff saved.
<img width="1057" alt="Captura de Tela 2023-06-24 às 15 36 13" src="https://github.com/azavea/raster-vision/assets/8769816/0c6ce47d-27ef-4143-a519-a3c3e4d60a7d">

What can be the problem here?

**Comment by AdeelH** [comment]
I'm not sure what I'm looking at here. As a sanity check, you can try predicting on the an image from the training or validation set. You can also [inspect the predictions](https://docs.rastervision.io/en/0.20/usage/tutorials/pred_and_eval_ss.html#Visualize-predictions) using `pred_labels.get_label_arr()` or `pred_labels.get_score_arr()`, to see if they look okay before saving them.

**Reply by lcoandrade** [reply]
I've just tested with `pred_labels.get_label_arr()` and `pred_labels.get_score_arr()`. As I have 13 classes, I'm more interested in the  `pred_labels.get_label_arr `.
I've tested with my validation dataset and the outputs are quite different from `learner.plot_predictions(split='valid', show=True)`. Not sure why.

I've tried with `raw_out=False` and `smooth=False` to see what would happen with the outputs.

Here is the window with the image file:
<img width="746" alt="Captura de Tela 2023-06-26 às 10 08 22" src="https://github.com/azavea/raster-vision/assets/8769816/469e67ae-a014-4e75-bb9e-0ce109375476">

Here is the labels used to train:
<img width="746" alt="Captura de Tela 2023-06-26 às 10 08 35" src="https://github.com/azavea/raster-vision/assets/8769816/7d36194d-ecac-4c93-88cf-f533c66636d2">

And here is the geo tiff file saved over the image:
<img width="750" alt="Captura de Tela 2023-06-26 às 10 17 20" src="https://github.com/azavea/raster-vision/assets/8769816/6104ae81-865b-49a8-b22c-efa779244d05">


These are some predicted values with just one epoch:
<img width="917" alt="Captura de Tela 2023-06-26 às 10 11 52" src="https://github.com/azavea/raster-vision/assets/8769816/4e719397-2849-48fc-9a73-caed34551811">

Maybe I'm missing something silly here, but I haven't identified the source of the problem yet....

Could you take a look at my [notebook](https://www.kaggle.com/code/luizclaudioandrade/rastervision-101?scriptVersionId=134925212&cellId=27) to see what should be changed?

Thanks in advance!

**Reply by AdeelH** [reply, marked as answer]
It seems like you're re-instantiating the `Learner` just before you make predictions. That means you're essentially predicting with a randomly initialized model.

- You do not need to re-instantiate the `Learner` before making predictions if you already have it in memory.
- If you do need to load a `Learner` from a model-bundle, use `Learner.from_model_bundle()`. See [this](https://docs.rastervision.io/en/0.20/usage/tutorials/train.html#For-predictions----Learner.from_model_bundle()) or [this](https://docs.rastervision.io/en/0.20/usage/tutorials/pred_and_eval_ss.html#Load-a-Learner-with-a-trained-model-from-bundle----Learner.from_model_bundle()).

**Reply by lcoandrade** [reply]
You are completely right!!!!

When I instantiate the learner like this:
```
learner = SemanticSegmentationLearner(
    cfg=learner_cfg,
    output_dir=OUTPUT_DIR,
    #model=model, # Not passing a model because I'm have a model config in learner_cfg
    train_ds=concat_dataset,
    valid_ds=validation_dataset,
)
```

When I pass the output_dir with a model bundle saved, it loads correctly and resumes from the epoch it has stopped earlier. So, that's why I tried to load the learner in prediction mode (training=False) in the same way.

I should have seem that! Thanks and sorry to bother!

**Reply by AdeelH** [reply]
> When I pass the output_dir with a model bundle saved, it loads correctly and resumes from the epoch it has stopped earlier. So, that's why I tried to load the learner in prediction mode (training=False) in the same way.

That's understandable. It only [does that](https://github.com/azavea/raster-vision/blob/4da111894d93ece46b157b08ffb3e0561d58e4a1/rastervision_pytorch_learner/rastervision/pytorch_learner/learner.py#L238) if `training=True`. And even in that case, it is not loading the state from the model bundle but rather the model file saved in the output dir.

>I should have seem that! Thanks and sorry to bother!

No problem! Feel free to ask more questions!

**Comment by lcoandrade** [comment]
> That's understandable. It only [does that](https://github.com/azavea/raster-vision/blob/4da111894d93ece46b157b08ffb3e0561d58e4a1/rastervision_pytorch_learner/rastervision/pytorch_learner/learner.py#L238) if `training=True`. And even in that case, it is not loading the state from the model bundle but rather the model file saved in the output dir.

Thanks! Now, I realize that.
